### ISSUE TITLE
Proper way to install docker compose (v2) ?

### ISSUE THREAD

**Issue by willcodeforfood** [vmware/photon#1487]
### Describe the bug

Running `docker compose` yields an error: `docker: 'compose' is not a docker command.`

There doesn't seem to be a recent guide or "best practice" recommendation for how to install/use `docker compose` on Photon. Much of what I find is outdated - for the old `docker-compose` (v1) tool. Or involves manually downloading it (without any possibility for auto-update from a repo)

Unless I've done something wrong, the scope of this question is more about "what's the best way to do this?" than "make it work".

### Reproduction steps

1. Run `tdnf install docker-compose` - presumably installs using [the SPEC file here](https://github.com/vmware/photon/blob/master/SPECS/docker-compose/docker-compose.spec). The setup completed with `Installing/Updating: docker-compose-2.14.0-3.ph5.x86_64`
2. Run `docker compose` - see error `docker: 'compose' is not a docker command.`
3. Run `docker plugin list` - see the list is empty

### Expected behavior

Running  `tdnf install docker-compose` should properly configure `docker compose` to where I can launch a container from a compose file.

### Additional context

I just updated this Photon instance to v5 (was previously v3). Previously I was using the `docker-compose` (v1) tool. This is my first time trying to use the newer `docker compose` cli on this Photon. It's possible I did something wrong - though I always remember it being very manual to install docker compose.
Hoping this GH "Issue" can serve as learning for others who may be in the same situation.

**Comment by dcasota** [comment]
Hi,

The docker-compose package in Ph5 installs `/bin/docker-compose`, actually with a version output `Docker Compose version 2.14.0`.

This isn't the same command as `docker compose`, indeed. Accordingly to https://github.com/docker/compose, _For a smooth transition from legacy docker-compose 1.xx, please consider installing [compose-switch](https://github.com/docker/compose-switch) to translate docker-compose ... commands into Compose V2's docker compose .... . Also check V2's --compatibility flag.

Please consider that there were Photon OS team voices, for example see https://github.com/vmware/photon/issues/647, which weren't enthusiasts about docker compose at that time.

In addition, the docs in mind, an upgrade from Photon OS 3.0 to 5.0 is not officially supported (yet?).

For Photon OS 4.0, the upgrade workflow is described in https://vmware.github.io/photon/docs-v5/installation-guide/upgrading-to-photon-os-4/. Step 2 says, _Stop any services (for example, docker) that are currently running in the VM._ and _After the upgrade, before you deploy into production, test all previous functionality to ensure that everything works as expected._.

Hence, with respect to "what's the best way to do this?", I would consider to assess your containers that are affected by the upgrade, customize a solution and verify that the controls function properly.
Beside that, imho it is good to have reached out to the Photon OS team. Usually if there are enough "voices", they do consider enhancements, but it also can take a few weeks before you get a qualified answer.

Hope this helps.
-Daniel

**Comment by willcodeforfood** [comment]
Thanks Daniel! Very helpful indeed.
Would it be correct to say that there's no "official" way to install `docker compose` on Photon at this time? I'm fine to install it as a one-off binary, but always try to avoid "veering off the beaten path" when there's a vendor-endorsed option.

Realizing "best" is highly subjective, to me it means: well-supported by the software vendor, such that I can automatically receive upstream updates as they release, and those updates have already been tested by the vendor/community to be compatible. Example: Ubuntu (and other distros) maintain their own repos of common packages, albeit with some lag-behind on releases to ensure compatibility.

If the Photon team is against doing something like that with `docker compose`, I can't fault them for that. I had assumed that they were [incorrectly, it seems], so I was looking for this "supported" solution.

In light of this, and given that I'm just a homelabber/VMWare enthusiast in a pseudo-production environment, it sounds like my best option is to install `docker compose` directly from Docker's own repo. Would you agree with that sentiment?

**Comment by dafanasiev** [comment]
Any news?

**Comment by dcasota** [comment]
Sure. Simply install docker. For `docker-compose` I'm using the following code snippet from the docs.
```
DOCKER_CONFIG=${DOCKER_CONFIG:-$HOME/.docker}
mkdir -p $DOCKER_CONFIG/cli-plugins
curl -SL https://github.com/docker/compose/releases/download/v2.27.1/docker-compose-linux-x86_64 -o $DOCKER_CONFIG/cli-plugins/docker-compose
sudo chmod a+x $DOCKER_CONFIG/cli-plugins/docker-compose
```


![image](https://github.com/vmware/photon/assets/14890243/ec03d05e-2730-4e07-a2e3-c8a199744037)

**Comment by dafanasiev** [comment]
@dcasota I think that "proper way" means "use tdnf (because its rpm-based os), not hands"

**Comment by dcasota** [comment]
Yes, it‘s really nice that tdnf is built-in in the internet versions of Photon OS. Same security awareness topics, robust, dependencies check for updates/upgrades, etc.
All latest&greatest bits from the internet but inter-pipeline-less are primarily to be taxed unsecure. Hence, there is a gap of 0-2 years to package managed versions.
Luckily, often, like in this case with docker-compose from the docs, the installation works flawlessly. 🍀

**Comment by WojciechThomas** [comment]
Why not:
```sh
sudo tdnf install  -y docker-compose
mkdir -p ~/.docker/cli-plugins
ln -s /usr/bin/docker-compose ~/.docker/cli-plugins/docker-compose
docker compose version
```

**Comment by dcasota** [comment]
Hi, actually there seems a package gap to docker compose [latest](https://github.com/docker/compose/releases) but as soon as it is a pretested automatic version bump for the open-source PhotonOS, yes, ‚why not’ is better.

**Comment by dcasota** [comment]
@EricSeastrand Good news: this now works out of the box on current Photon OS 5.0. `sudo tdnf install docker-compose` installs docker-compose-5.1.4-1.ph5, and the package ships the Compose V2 binary as a proper Docker CLI plugin at /usr/libexec/docker/cli-plugins/docker-compose (in addition to the standalone /usr/bin/docker-compose). Since /usr/libexec/docker/cli-plugins is one of Docker's default plugin search paths, `docker compose version` works immediately, no manual symlink or download needed. So the tdnf-managed, vendor-supported path you were asking for now exists (just be aware the packaged version lags upstream Compose by a bit). Could you confirm this covers your need so we can close this? If you'd still prefer the newest upstream release, the manual ~/.docker/cli-plugins install from the Docker docs remains an option.

**Comment by EricSeastrand** [comment]
Thanks @dcasota! Happy to hear this package is now in the official repo! I will mark the issue as closed.

I must admit: I have not used Photon since the Broadcom acquisition, so I can't test this directly. But I'll be glad it's there for the next time I pick up Photon.
